lib/identity_cache/cached/prefetcher.rb

```diff
@@ -0,0 +1,200 @@
+          unless records.first.class.should_use_cache?
+            ActiveRecord::Associations::Preloader.new.preload(records, association)
+            return
+          end
+
+          if (cached_has_many = klass.cached_has_manys[association])
+            fetch_has_many(cached_has_many, records)
+          elsif (cached_belongs_to = klass.cached_belongs_tos[association])
+            fetch_belongs_to(cached_belongs_to, records)
+          elsif (cached_has_one = klass.cached_has_ones[association])
+            fetch_has_one(cached_has_one, records)
+          else
+            raise ArgumentError, "Unknown cached association #{association} listed for prefetching"
+          end
+        end
+
+        def fetch_has_many(cached_has_many, records)
+          prefetch_embedded(cached_has_many, records)
+
+          if cached_has_many.embedded_recursively?
+            child_records = records.flat_map(&cached_has_many.cached_accessor_name.to_sym)
+          else
+            ids_to_parent_record = records.each_with_object({}) do |record, hash|
+              child_ids = record.send(cached_has_many.cached_ids_name)
+              child_ids.each do |child_id|
+                hash[child_id] = record
+              end
+            end
+
+            parent_record_to_child_records = Hash.new { |h, k| h[k] = [] }
+            child_records = cached_has_many.reflection.klass.fetch_multi(*ids_to_parent_record.keys)
+            child_records.each do |child_record|
+              parent_record = ids_to_parent_record[child_record.id]
+              parent_record_to_child_records[parent_record] << child_record
+            end
+
+            parent_record_to_child_records.each do |parent, children|
+              parent.instance_variable_set(cached_has_many.records_variable_name, children)
+            end
+          end
+
+          child_records
+        end
+
+        def fetch_has_one(cached_has_one, records)
+          if cached_has_one.embedded_recursively?
+            prefetch_embedded(cached_has_one, records)
+            parent_records = records.map(&cached_has_one.cached_accessor_name.to_sym).compact
+          else
+            ids_to_parent_record = records.each_with_object({}) do |record, hash|
+              child_id = record.send(cached_has_one.cached_id_name)
+              hash[child_id] = record if child_id
+            end
+
+            parent_record_to_child_record = {}
+            child_records = cached_has_one.reflection.klass.fetch_multi(*ids_to_parent_record.keys)
+            child_records.each do |child_record|
+              parent_record = ids_to_parent_record[child_record.id]
+              parent_record_to_child_record[parent_record] ||= child_record
+            end
+
+            parent_record_to_child_record.each do |parent, child|
+              parent.instance_variable_set(cached_has_one.records_variable_name, child)
+            end
+          end
+
+          parent_records
```

Comment:
I think we might have the naming backwards here and the associated records should be referred to as child records rather than parent records.  Looks like that was already the case with the code on master though, so I guess changing it is out of scope of this PR.

If we want a generic term to refer to them across types of associations, then we could refer to them as associated records in the future.

Reply:
Sure, I'll address this in another PR